source/llassetgen/include/llassetgen/Packing.h

```diff
@@ -0,0 +1,176 @@
+#pragma once
+
+#include <algorithm>
+#include <cstdint>
+#include <iterator>
+#include <type_traits>
+#include <vector>
+
+#include <llassetgen/Geometry.h>
+#include <llassetgen/llassetgen_api.h>
+
+namespace llassetgen {
+    using PackingSizeType = unsigned int;
+
+    /*
+     * Describes the packing of a texture atlas.
+     *
+     * The rectangles in `rects` correspond to the input rectangles at the same
+     * position in the packing algorithms input.
+     */
+    struct Packing {
+        Vec2<PackingSizeType> atlasSize{};
+        std::vector<Rect<PackingSizeType>> rects{};
+
+        Packing() = default;
+
+        LLASSETGEN_API explicit Packing(Vec2<PackingSizeType> _atlasSize);
+    };
+
+    namespace impl_packing {
+        template <class Iter>
+        constexpr int assertCorrectIteratorType() {
+            using IterTraits = typename std::iterator_traits<Iter>;
+            using IterRefType = typename IterTraits::reference;
+            using IterCategory = typename IterTraits::iterator_category;
+            static_assert(std::is_convertible<IterRefType, const Vec2<PackingSizeType>&>::value,
+                          "Packing input must be convertible to const Vec2<PackingSizeType>&");
+            static_assert(std::is_base_of<std::forward_iterator_tag, IterCategory>::value,
+                          "Iterator must be a ForwardIterator");
+
+            // C++11 doesn't allow void as the return type of constexpr functions
```

Comment:
Why even wrapping it in a function? `static_assert` is evaluated in class contexts, too.

Reply:
I factored this out of the two `shelfPack*` functions to avoid code duplication. Which class context would you move this too? I tried to keep `ShelfNextFitPacker` a non-templated class so that this doesn't become a header-only part of an otherwise not header-only library.

Also, having a function constraints the type aliases to where they are needed.